cli/import_map.rs

```diff
@@ -0,0 +1,640 @@
+
+    for address in potential_addresses.iter() {
+      let potential_address = match address {
+        Value::String(address) => address,
+        _ => continue,
+      };
+      match resolve_module_spec(potential_address, base_url) {
+        Ok(address_url) => {
+          if specifier_key.ends_with("/") && !address_url.ends_with("/") {
+            println!(
+              "Invalid target address `{:?}` for package specifier `{:?}`.\
+               Package address targets must end with `/`.",
+              address_url, specifier_key
+            );
+            continue;
+          }
+
+          normalized_addresses.push(address_url);
+        }
+        _ => continue,
+      }
+    }
+
+    normalized_addresses
+  }
+
+  fn normalize_specifier_map(
+    imports_map: &Map<String, Value>,
+    base_url: &str,
+  ) -> HashMap<String, Vec<String>> {
+    let mut normalized_map: HashMap<String, Vec<String>> = HashMap::new();
+
+    for (specifier_key, value) in imports_map.iter() {
+      let normalized_specifier_key =
+        match ImportMap::normalize_specifier_key(specifier_key, base_url) {
+          Some(s) => s,
+          None => continue,
+        };
+
+      let normalized_address_array = ImportMap::normalize_addresses(
+        &normalized_specifier_key,
+        base_url,
+        value.clone(),
+      );
+
+      normalized_map.insert(normalized_specifier_key, normalized_address_array);
+    }
+
+    normalized_map
+  }
+
+  pub fn from_json(
+    base_url: &str,
+    json_string: &str,
+  ) -> Result<Self, ImportMapError> {
+    let v: Value = match serde_json::from_str(json_string) {
+      Ok(v) => v,
+      Err(_) => {
+        return Err(ImportMapError::new("Unable to parse import map JSON"));
+      }
+    };
+
+    match v {
+      Value::Object(_) => {}
+      _ => {
+        return Err(ImportMapError::new("Import map JSON must be an object"));
+      }
+    }
+
+    let normalized_imports = match &v["imports"] {
+      Value::Object(imports_map) => {
+        ImportMap::normalize_specifier_map(imports_map, base_url)
+      }
+      _ => {
+        return Err(ImportMapError::new(
+          "Import map's 'imports' must be an object",
+        ));
+      }
+    };
+
+    // TODO(bartlomieju): handle scopes
+
+    let import_map = ImportMap {
+      base_url: base_url.to_string(),
+      modules: normalized_imports,
+    };
+
+    Ok(import_map)
+  }
+
+  // TODO(bartlomieju): this method can definitely be optimized
+  // https://github.com/WICG/import-maps/issues/73#issuecomment-439327758
+  // for some candidate implementations.
+
+  // TODO(bartlomieju):
+  // "most-specific wins", i.e. when there are multiple matching keys,
+  // choose the longest.
+  // https://github.com/WICG/import-maps/issues/102
+  fn resolve_imports_match(
+    &self,
+    normalized_specifier: String,
+  ) -> Option<String> {
+    for (specifier_key, address_vec) in self.modules.iter() {
```

Comment:
During this iteration we should sort `self.modules` by longest keys (for most specific match), reference:

https://github.com/WICG/import-maps/blob/faf0ea06f5ac2a29bd1cd7734951e0c0a035adc2/reference-implementation/lib/parser.js#L87

Or probably it'd be better to sort them during parsing and iterate over with insertion order.